server/historian/packages/historian-base/src/utils.ts

```diff
@@ -3,18 +3,29 @@
  * Licensed under the MIT License.
  */
 
+import { ITokenClaims } from "@fluidframework/protocol-definitions";
+import * as jwt from "jsonwebtoken";
+
 export function normalizePort(val) {
     const normalizedPort = parseInt(val, 10);
 
     if (isNaN(normalizedPort)) {
-    // named pipe
+        // named pipe
         return val;
     }
 
     if (normalizedPort >= 0) {
-    // port number
+        // port number
         return normalizedPort;
     }
 
     return false;
 }
+
+export function getTokenLifetimeInSec(token: string): number {
+    const claims = jwt.decode(token) as ITokenClaims;
+    if (claims && claims.exp) {
+        return ((claims.exp - Math.round((new Date()).getTime())) * 1000);
+    }
+    return undefined;
+}
```

Comment:
This is an interesting problem. Imagine the clock of this service is behind by 2 minutes. So we are actually caching it for 62 minutes. For the last 2 minutes, we are still using a stale token even if the client sent us a fresh new token. Its may not be a big deal but I am not a security expert either. So may be we should reduce the caching time by 5%?